File: ios/Runner/BatteryChannel.swift
import Flutter
import Foundation
import UIKit
import os.log

/// Platform channel handler for battery information on iOS.
///
/// Provides battery level and charging state to the Dart side via
/// FlutterMethodChannel "com.redgrid.link/battery".
///
/// Used by Field Link to include battery info in peer status broadcasts
/// and to enable battery-aware expedition mode.
class BatteryChannel: NSObject {

    // MARK: - Constants

    private static let methodChannelName = "com.redgrid.link/battery"

    private let log = OSLog(subsystem: "com.redgrid.link", category: "BatteryChannel")

    // MARK: - Flutter channel

    private var methodChannel: FlutterMethodChannel?

    // MARK: - Registration

    func register(with messenger: FlutterBinaryMessenger) {
        // Enable battery monitoring so UIDevice reports meaningful values
        UIDevice.current.isBatteryMonitoringEnabled = true

        methodChannel = FlutterMethodChannel(
            name: BatteryChannel.methodChannelName,
            binaryMessenger: messenger
        )
        methodChannel?.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result: result)
        }

        os_log("BatteryChannel registered", log: log, type: .info)
    }

    func unregister() {
        methodChannel?.setMethodCallHandler(nil)
        methodChannel = nil
    }

    // MARK: - Method dispatch

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getBatteryLevel":
            getBatteryLevel(result: result)
        case "getBatteryState":
            getBatteryState(result: result)
        case "getBatteryInfo":
            getBatteryInfo(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Battery level

    /// Returns the current battery level as an integer percentage (0-100).
    /// Returns -1 if the battery level cannot be determined (e.g., on Simulator).
    private func getBatteryLevel(result: @escaping FlutterResult) {
        let level = UIDevice.current.batteryLevel
        if level < 0 {
            // -1.0 means battery level is unknown (e.g., Simulator)
            result(-1)
        } else {
            result(Int(level * 100))
        }
    }

    // MARK: - Battery state

    /// Returns the current battery charging state as a string:
    /// - "charging"
    /// - "discharging" (mapped from .unplugged)
    /// - "full"
    /// - "unknown"
    private func getBatteryState(result: @escaping FlutterResult) {
        let state = UIDevice.current.batteryState
        result(batteryStateString(state))
    }

    private func batteryStateString(_ state: UIDevice.BatteryState) -> String {
        switch state {
        case .charging:
            return "charging"
        case .unplugged:
            return "discharging"
        case .full:
            return "full"
        case .unknown:
            return "unknown"
        @unknown default:
            return "unknown"
        }
    }

    // MARK: - Combined battery info

    /// Returns a map with both level and state for efficiency.
    private func getBatteryInfo(result: @escaping FlutterResult) {
        let level = UIDevice.current.batteryLevel
        let state = UIDevice.current.batteryState

        let levelInt = level < 0 ? -1 : Int(level * 100)

        result([
            "level": levelInt,
            "state": batteryStateString(state),
            // iOS doesn't expose battery temperature via public APIs
            "temperature": -1.0,
        ] as [String: Any])
    }
}

File: android/app/src/main/kotlin/com/redgrid/red_grid_link/BatteryChannel.kt
package com.redgrid.red_grid_link

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build
import android.util.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Platform channel handler for battery information.
 *
 * Provides battery level and charging state to the Dart side via
 * MethodChannel "com.redgrid.link/battery".
 *
 * Used by Field Link to include battery info in peer status broadcasts
 * and to enable battery-aware expedition mode.
 */
class BatteryChannel(
    private val context: Context,
) : MethodChannel.MethodCallHandler {

    companion object {
        private const val TAG = "BatteryChannel"
        private const val METHOD_CHANNEL = "com.redgrid.link/battery"
    }

    private lateinit var methodChannel: MethodChannel

    // -------------------------------------------------------------------------
    // Registration
    // -------------------------------------------------------------------------

    fun register(flutterEngine: FlutterEngine) {
        methodChannel = MethodChannel(
            flutterEngine.dartExecutor.binaryMessenger,
            METHOD_CHANNEL,
        )
        methodChannel.setMethodCallHandler(this)
        Log.d(TAG, "BatteryChannel registered")
    }

    fun unregister() {
        methodChannel.setMethodCallHandler(null)
    }

    // -------------------------------------------------------------------------
    // MethodChannel dispatch
    // -------------------------------------------------------------------------

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getBatteryLevel" -> getBatteryLevel(result)
            "getBatteryState" -> getBatteryState(result)
            "getBatteryInfo" -> getBatteryInfo(result)
            else -> result.notImplemented()
        }
    }

    // -------------------------------------------------------------------------
    // Battery level
    // -------------------------------------------------------------------------

    /**
     * Returns the current battery level as an integer percentage (0-100).
     * Returns -1 if the battery level cannot be determined.
     */
    private fun getBatteryLevel(result: MethodChannel.Result) {
        try {
            val level = getBatteryLevelInternal()
            result.success(level)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get battery level", e)
            result.error(
                "BATTERY_ERROR",
                "Failed to get battery level: ${e.message}",
                null,
            )
        }
    }

    private fun getBatteryLevelInternal(): Int {
        val batteryManager = context.getSystemService(Context.BATTERY_SERVICE) as? BatteryManager
        if (batteryManager != null) {
            return batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
        }

        // Fallback: read from sticky broadcast
        val intent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        if (intent != null) {
            val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
            val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
            if (level >= 0 && scale > 0) {
                return (level * 100) / scale
            }
        }

        return -1
    }

    // -------------------------------------------------------------------------
    // Battery state
    // -------------------------------------------------------------------------

    /**
     * Returns the current battery charging state as a string:
     * - "charging"
     * - "discharging"
     * - "full"
     * - "not_charging"
     * - "unknown"
     */
    private fun getBatteryState(result: MethodChannel.Result) {
        try {
            val state = getBatteryStateInternal()
            result.success(state)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get battery state", e)
            result.error(
                "BATTERY_ERROR",
                "Failed to get battery state: ${e.message}",
                null,
            )
        }
    }

    private fun getBatteryStateInternal(): String {
        val intent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
        if (intent == null) return "unknown"

        val status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
        return when (status) {
            BatteryManager.BATTERY_STATUS_CHARGING -> "charging"
            BatteryManager.BATTERY_STATUS_DISCHARGING -> "discharging"
            BatteryManager.BATTERY_STATUS_FULL -> "full"
            BatteryManager.BATTERY_STATUS_NOT_CHARGING -> "not_charging"
            else -> "unknown"
        }
    }

    // -------------------------------------------------------------------------
    // Combined battery info
    // -------------------------------------------------------------------------

    /**
     * Returns a map with both level and state for efficiency
     * (single call from Dart instead of two).
     */
    private fun getBatteryInfo(result: MethodChannel.Result) {
        try {
            val level = getBatteryLevelInternal()
            val state = getBatteryStateInternal()

            val intent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
            val temperature = if (intent != null) {
                intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, -1) / 10.0
            } else {
                -1.0
            }

            result.success(mapOf(
                "level" to level,
                "state" to state,
                "temperature" to temperature,
            ))
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get battery info", e)
            result.error(
                "BATTERY_ERROR",
                "Failed to get battery info: ${e.message}",
                null,
            )
        }
    }
}
